### DISCUSSION TITLE
Can I use SB with the Page Object Model?

### DISCUSSION THREAD

**Discussion by drdedou** [seleniumbase/SeleniumBase#2494] [Q&A] [answered]
from seleniumbase import SB

with SB(uc=True) as sb:

**Comment by mdmintz** [comment, marked as answer]
The official Page Object Model formats with SeleniumBase are:

[Syntax Format 5](https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/syntax_formats.md#sb_sf_05) - with the `BaseCase` format:

```python
from seleniumbase import BaseCase
BaseCase.main(__name__, __file__)

class LoginPage:
    def login_to_swag_labs(self, sb, username):
        sb.open("https://www.saucedemo.com")
        sb.type("#user-name", username)
        sb.type("#password", "secret_sauce")
        sb.click('input[type="submit"]')

class MyTests(BaseCase):
    def test_swag_labs_login(self):
        LoginPage().login_to_swag_labs(self, "standard_user")
        self.assert_element("div.inventory_list")
        self.assert_element('div:contains("Sauce Labs Backpack")')
```

and [Syntax Format 6](https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/syntax_formats.md#sb_sf_06) with the `pytest` fixture:

```python
class LoginPage:
    def login_to_swag_labs(self, sb, username):
        sb.open("https://www.saucedemo.com")
        sb.type("#user-name", username)
        sb.type("#password", "secret_sauce")
        sb.click('input[type="submit"]')

class MyTests:
    def test_swag_labs_login(self, sb):
        LoginPage().login_to_swag_labs(sb, "standard_user")
        sb.assert_element("div.inventory_list")
        sb.assert_element('div:contains("Sauce Labs Backpack")')
```

**Comment by drdedou** [comment]
how i can use [Syntax Format 5] on Jupyter i get 
NameError: name '__file__' is not defined

**Comment by drdedou** [comment]
and this 

OutOfScopeException                       Traceback (most recent call last)
[~\AppData\Local\Temp\ipykernel_12060\3747152559.py](https://file+.vscode-resource.vscode-cdn.net/f%3A/Project/Automation/Python%20Selenium/BlsAutomation/seleniumbase/Project/~/AppData/Local/Temp/ipykernel_12060/3747152559.py) in ?()
      1 sbc = BaseCase()
      2 
      3 Url = "https://algeria.blsspainglobal.com/dza/home/index"
----> 4 sbc.get_new_driver(undetectable=True)
      5 sbc.driver.uc_open_with_reconnect(
      6     Url, reconnect_time=3
      7 )

[e:\programme\anaconda3\envs\seleniumbase\Lib\site-packages\seleniumbase\fixtures\base_case.py](file:///E:/programme/anaconda3/envs/seleniumbase/Lib/site-packages/seleniumbase/fixtures/base_case.py) in ?(self, browser, headless, locale_code, protocol, servername, port, proxy, proxy_bypass_list, proxy_pac_url, multi_proxy, agent, switch_to, cap_file, cap_string, recorder_ext, disable_js, disable_csp, enable_ws, enable_sync, use_auto_ext, undetectable, uc_cdp_events, uc_subprocess, log_cdp_events, no_sandbox, disable_gpu, headless2, incognito, guest_mode, dark_mode, devtools, remote_debug, enable_3d_apis, swiftshader, ad_block_on, host_resolver_rules, block_images, do_not_track, chromium_arg, firefox_arg, firefox_pref, user_data_dir, extension_zip, extension_dir, disable_features, binary_location, driver_version, page_load_strategy, use_wire, external_pdf, is_mobile, d_width, d_height, d_p_r)
   3846         d_width - the device width of the mobile emulator (Chrome-only)
   3847         d_height - the device height of the mobile emulator (Chrome-only)
   3848         d_p_r - the device pixel ratio of the mobile emulator (Chrome-only)
   3849         """
-> 3850         self.__check_scope()
   3851         if self.browser == "remote" and self.servername == "localhost":
   3852             raise Exception(
   3853                 'Cannot use "remote" browser driver on localhost!'

[e:\programme\anaconda3\envs\seleniumbase\Lib\site-packages\seleniumbase\fixtures\base_case.py](file:///E:/programme/anaconda3/envs/seleniumbase/Lib/site-packages/seleniumbase/fixtures/base_case.py) in ?(self)
  10501                 "\n they can call BaseCase class methods with all the required"
  10502                 "\n variables, which are initialized during the setUp() method"
  10503                 "\n that runs automatically before all tests called by pytest."
...
 When using page objects, be sure to pass the `self` object
 from your test class into your page object methods so that
 they can call BaseCase class methods with all the required
 variables, which are initialized during the setUp() method
 that runs automatically before all tests called by pytest.

**Comment by mdmintz** [comment]
Use the syntax exactly as written in https://github.com/seleniumbase/SeleniumBase/discussions/2494#discussioncomment-8480114

**Comment by drdedou** [comment]
![Screenshot 2024-02-15 151447](https://github.com/seleniumbase/SeleniumBase/assets/47856801/2e3fea1c-d795-422e-8311-43039c823eff)

yes i uses the same

**Reply by mdmintz** [reply]
Run the script with `pytest`, not `python`.

If you're not calling a file directly from the command-line, remove the `BaseCase.main(__name__, __file__)` line.
